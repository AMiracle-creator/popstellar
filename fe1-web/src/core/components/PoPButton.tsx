import PropTypes from 'prop-types';
import React from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';

import { Border, Color, Spacing } from '../styles';
import PoPTouchableOpacity from './PoPTouchableOpacity';

const styles = StyleSheet.create({
  container: {
    marginBottom: Spacing.x1,
  } as ViewStyle,
  button: {
    padding: Spacing.x05,
    borderColor: Color.accent,
    backgroundColor: Color.accent,
    borderWidth: Border.width,
    borderRadius: Border.radius,
  } as ViewStyle,
  disabled: {
    borderColor: Color.inactive,
    backgroundColor: Color.inactive,
  } as ViewStyle,
  outline: {
    borderColor: Color.accent,
    backgroundColor: Color.contrast,
  } as ViewStyle,
  negativeBorder: {
    borderColor: Color.contrast,
  } as ViewStyle,
  disabledNegativeBorder: {
    borderColor: Color.inactive,
    backgroundColor: Color.inactive,
  } as ViewStyle,
});

const PoPButton = (props: IPropTypes) => {
  const { onPress, disabled, children, outline, negativeBorder, margin, testID } = props;

  const viewStyles = [styles.button];
  if (negativeBorder && disabled) {
    viewStyles.push(styles.disabledNegativeBorder);
  } else if (negativeBorder) {
    viewStyles.push(styles.negativeBorder);
  } else if (disabled) {
    viewStyles.push(styles.disabled);
  } else if (outline) {
    viewStyles.push(styles.outline);
  }

  return (
    <PoPTouchableOpacity
      containerStyle={margin ? styles.container : []}
      onPress={disabled ? undefined : onPress}
      testID={testID || undefined}>
      <View style={viewStyles}>{children}</View>
    </PoPTouchableOpacity>
  );
};

const propTypes = {
  onPress: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  outline: PropTypes.bool,
  negativeBorder: PropTypes.bool,
  margin: PropTypes.bool,
  children: PropTypes.node,
  testID: PropTypes.string,
};
PoPButton.propTypes = propTypes;

PoPButton.defaultProps = {
  disabled: false,
  outline: false,
  negativeBorder: false,
  margin: true,
  children: null,
  testID: undefined,
};

type IPropTypes = PropTypes.InferProps<typeof propTypes>;

export default PoPButton;
