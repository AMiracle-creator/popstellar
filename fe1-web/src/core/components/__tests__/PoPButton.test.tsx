import { render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';

import PoPButton from '../PoPButton';

let onPress: Function;
const buttonText = 'I am a wide button';

beforeEach(() => {
  onPress = jest.fn();
});

describe('PoPButton', () => {
  it('renders correctly without disabled', () => {
    const { toJSON } = render(
      <PoPButton onPress={() => onPress}>
        <Text>{buttonText}</Text>
      </PoPButton>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
