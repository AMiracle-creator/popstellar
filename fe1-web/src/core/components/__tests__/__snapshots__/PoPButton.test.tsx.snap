// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoPButton renders correctly without disabled 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#3742fa",
          "borderColor": "#3742fa",
          "borderRadius": 8,
          "borderWidth": 1,
          "padding": 8,
        },
      ]
    }
  >
    <Text>
      I am a wide button
    </Text>
  </View>
</View>
`;
